pkg/controller/controller_instance.go

```diff
@@ -1198,6 +1198,12 @@ func (c *controller) recordStartOfServiceInstanceOperation(toUpdate *v1beta1.Ser
 		reason,
 		message,
 	)
+
+	// clear the polling queue, just in case the controller hadn't reset it
```

Comment:
To clarify: the `finishPollingServiceInstance` method invokes the `c.instancePollingQueue.Forget(key)` method, which has a following comment:
```
	// Forget indicates that an item is finished being retried.  Doesn't matter whether its for perm failing
	// or for success, we'll stop the rate limiter from tracking it.  This only clears the `rateLimiter`, you
	// still have to call `Done` on the queue.
```
and particularly this:
> This only clears the `rateLimiter`, you still have to call `Done` on the queue.

So your comment "clear the polling queue" is misleading. We just clear the `rateLimiter`, and the item remains in the queue.

Actually, the comment to `finishPollingServiceInstance` method is incorrect as well:
```
// finishPollingServiceInstance removes the instance's key from the controller's instance
// polling queue.
func (c *controller) finishPollingServiceInstance
```
so it would be nice to fix it too.

Reply:
Which makes me think that in some situations we might want to invoke the `Done` method instead of `Forget`? When we finished the processing I mean.

In the case `recordStartOfServiceInstanceOperation` invoking `Forget` seems correct.

Reply:
Do we need a method called something like "resetRateLimiter"?

Reply:
@nilebox 

For full context, the way the rate-limited queues work is that you call `Get` to pull an item out of the queue, and then call `Done` when you're finished with that pass over the item. This is the locking mechanism the queue uses to be thread-safe, ensuring it will never give multiple threads the same resource at the same time accidentally. If you want to add it back into the queue, you call `queue.AddRateLimited`, and if you want to clear the rate-limiter, you call `queue.Forget`. Note that `Done` is not a way to remove the resource from the queue; it only marks that you've finished your current pass over it, and `Done` will in fact requeue the resource if it has been marked dirty in the meantime.

Here, the calling of `Get` and `Done` are done by the workers. Every time they pull something off the queue, they call `Get`, and once they're finished with that pass, they call `Done`. In the case of the polling queue, the only thing they do in the "pass" is move the resource from the polling queue into the instance queue, which is nearly instantaneous.

As to the first comment, I agree that the comments are a bit misleading. It's a bit tricky to accurately describe what it's doing, though, because there are assumptions under which it's operating. In this case, we know that since the resource has been passed from the polling queue to the instance queue (and thus `Done` has already been called on that resource for the polling queue), and since it will never get requeued in the polling queue until `reconcileServiceInstance` calls `[begin/continue]PollingServiceInstance`, the only memory the polling queue would have of the resource is from its rate limiter, and so calling `Forget` will truthfully clear all artifacts the polling queue has of that instance. I can try to make the comments better if you still feel under these assumptions that they're inaccurate.

As to the second comment, as described above, `Done` is handled by the worker function, and we should never touch that as part of reconciliation.

Reply:
@kibbles-n-bytes thanks for the detailed info!
In the case of `recordStartOfServiceInstanceOperation` though invoking the `finishPollingServiceInstance`  method and removing item from the queue doesn't make sense - that's why I started checking what this method actually does, and what the comments say.

> In this case, we know that since the resource has been passed from the polling queue to the instance queue (and thus Done has already been called on that resource for the polling queue)

I don't think it's fair to refer to this assumption there. You don't care whether the item was removed from the polling queue, and whether your `Forget` call will remove it from the queue. You just want to reset the rate limiting delay before you start polling for the new operation.
So I would prefer to invoke a method called `resetPollingRateLimiter` or something similar to reflect its purpose.  

As of the other places - I'm fine with keeping them as they are, however I still think that the comments are still misleading as they unnecessarily refer to  the behavior of other parts of the code.  So it would probably nice to just rename the `finishPollingServiceInstance` method to `resetPollingRateLimiter` instead of creating a new method.